ports/gstplugin/servosrc.rs

```diff
@@ -0,0 +1,566 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use crate::logging::CATEGORY;
+
+use crossbeam_channel::Receiver;
+use crossbeam_channel::Sender;
+
+use euclid::Point2D;
+use euclid::Rect;
+use euclid::Scale;
+use euclid::Size2D;
+
+use glib::glib_bool_error;
+use glib::glib_object_impl;
+use glib::glib_object_subclass;
+use glib::object::Cast;
+use glib::subclass::object::ObjectImpl;
+use glib::subclass::object::ObjectImplExt;
+use glib::subclass::simple::ClassStruct;
+use glib::subclass::types::ObjectSubclass;
+use gstreamer::gst_element_error;
+use gstreamer::gst_loggable_error;
+use gstreamer::subclass::element::ElementClassSubclassExt;
+use gstreamer::subclass::element::ElementImpl;
+use gstreamer::subclass::ElementInstanceStruct;
+use gstreamer::BufferRef;
+use gstreamer::Caps;
+use gstreamer::CoreError;
+use gstreamer::ErrorMessage;
+use gstreamer::FlowError;
+use gstreamer::FlowSuccess;
+use gstreamer::Format;
+use gstreamer::Fraction;
+use gstreamer::FractionRange;
+use gstreamer::IntRange;
+use gstreamer::LoggableError;
+use gstreamer::PadDirection;
+use gstreamer::PadPresence;
+use gstreamer::PadTemplate;
+use gstreamer_base::subclass::base_src::BaseSrcImpl;
+use gstreamer_base::BaseSrc;
+use gstreamer_base::BaseSrcExt;
+use gstreamer_video::VideoFormat;
+use gstreamer_video::VideoFrameRef;
+use gstreamer_video::VideoInfo;
+
+use log::debug;
+use log::info;
+
+use servo::compositing::windowing::AnimationState;
+use servo::compositing::windowing::EmbedderCoordinates;
+use servo::compositing::windowing::EmbedderMethods;
+use servo::compositing::windowing::WindowEvent;
+use servo::compositing::windowing::WindowMethods;
+use servo::embedder_traits::EventLoopWaker;
+use servo::msg::constellation_msg::TopLevelBrowsingContextId;
+use servo::servo_url::ServoUrl;
+use servo::webrender_api::units::DevicePixel;
+use servo::Servo;
+
+use sparkle::gl;
+use sparkle::gl::types::GLuint;
+use sparkle::gl::Gl;
+
+use surfman::platform::generic::universal::context::Context;
+use surfman::platform::generic::universal::device::Device;
+use surfman::SurfaceAccess;
+use surfman::SurfaceType;
+
+use surfman_chains::SwapChain;
+
+use std::cell::RefCell;
+use std::rc::Rc;
+use std::sync::Mutex;
+use std::thread;
+
+pub struct ServoSrc {
+    sender: Sender<ServoSrcMsg>,
+    swap_chain: SwapChain,
+    info: Mutex<Option<VideoInfo>>,
+}
+
+struct ServoSrcGfx {
+    device: Device,
+    context: Context,
+    gl: Rc<Gl>,
+    fbo: GLuint,
+}
+
+impl ServoSrcGfx {
+    fn new() -> ServoSrcGfx {
+        let version = surfman::GLVersion { major: 4, minor: 3 };
+        let flags = surfman::ContextAttributeFlags::empty();
+        let attributes = surfman::ContextAttributes { version, flags };
+
+        let connection = surfman::Connection::new().expect("Failed to create connection");
```

Comment:
All this allocation of resources should ideally happen in `ElementImpl::change_state` in `gst::StateChange::NullToReady` and be deallocated again in `ReadyToNull`

Reply:
Or `BaseSrcImpl::start` and `::stop`

Reply:
Indeed, but you could start the gfx stuff (and shut down) at that point. Where is that happening currently, where would this code be triggered currently? :)

Reply:
I could do that, but it's not obvious that the benefit (deallocating on `stop` rather than at the program end) is worth the complexity (making the thread-local an option which refers to a flag saying whether or not the plugin has been started).

ATM the TLS is only allocated when it is used, so the servo thread allocates on startup and the gstreamer threads allocate the first time they call `fill`.

Reply:
Will that work if you try to use multiple servo sources at the same time in the same process?

Doing it globally seems fine too, I assume an application that uses the servo source once is going to not be bothered by the fact that for the whole runtime of the program there will be some servo thread running.

Reply:
Not really, what exactly are you looking for? :) You need to ensure that it's possible to have multiple instances at the same time and that they don't cause each other to break. If that's not possible at all by design (e.g. because exclusive hardware access), then there needs to be a mechanism implemented that the second instance gracefully fails. Ideally during the `NullToReady` state transition of the element.

Reply:
Well, docs about what is shared between the multiple instances of the same element would be nice. I'm guessing that the class will only be instantiated once, but each instance of the element will get it's own object?

Reply:
Yes, there's nothing shared between instances other than what you do in your code yourself. In terms of your code, using two `servosrc` would cause two separate `ServoSrc` to be created.

Reply:
OK, that will do for the moment, but it's a bit wasteful. Really we should be sharing servo state between the instances, and sending a `NewBrowser` message for each of them. But that can be a follow-up.

Reply:
Agreed :)

Reply:
Filed https://github.com/servo/servo/issues/24840